src/Compilers/Core/Portable/SourceGeneration/IncrementalContexts.cs

```diff
@@ -160,7 +162,15 @@ internal SourceProductionContext(AdditionalSourcesCollection sources, Diagnostic
         /// <remarks>
         /// The severity of the diagnostic may cause the compilation to fail, depending on the <see cref="Compilation"/> settings.
         /// </remarks>
-        public void ReportDiagnostic(Diagnostic diagnostic) => Diagnostics.Add(diagnostic);
+        /// <exception cref="ArgumentException">
+        /// <paramref name="diagnostic"/> is located in a syntax tree which is not part of the compilation,
+        /// its location span is outside of the given file, or its identifier is not valid.
+        /// </exception>
+        public void ReportDiagnostic(Diagnostic diagnostic)
+        {
+            DiagnosticAnalysisContextHelpers.VerifyArguments(diagnostic, Compilation, static (_, _) => true, CancellationToken);
```

Comment:
```suggestion
            DiagnosticAnalysisContextHelpers.VerifyArguments(diagnostic, Compilation, static (_, _) => true, CancellationToken);
```

Can you turn the lambda into a named argument as it's not clear at a glance what this is doing.